### PR TITLE
Extend SubsetSpec API

### LINKED ISSUE #5024: Support Envoy's default_subset and fallback_policy in Gloo's SubsetSpec
**Is your feature request related to a problem? Please describe.**

At the moment Gloo uses Envoy's ANY_ENDPOINT fallback_policy, which leads to requests that have a subset destination with empty subset being routed to any of the hosts in the upstream. However, sometimes this is not desirable and you'd prefer to route to a specific default subset instead. This is supported by Envoy by setting the fallback_policy option to DEFAULT_SUBSET and specifying selectors that define a default subset with the default_subset option. 

Reference: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/subsets

**Describe the solution you'd like**

I'd like to extend the Gloo [SubsetSpec](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/options/subset_spec.proto.sk/#subsetspec) to include the two fields
- fallbackPolicy: string
- defaultSubset: []options.gloo.solo.io.Selector

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by jannes**
# Description
This closes #5024

extend [options.gloo.solo.io.SubsetSpec](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/options/subset_spec.proto.sk/#subsetspec) with:

- fallbackPolicy: string
- defaultSubset: []options.gloo.solo.io.Selector

extend [options.gloo.solo.io.Selector](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/options/subset_spec.proto.sk/#selector) with:
- single_host_per_subset: boolean

These changes just expose already existing options of the Envoy API.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/5024

### CODE REVIEW

File: projects/gloo/pkg/translator/clusters.go [hunk 1]
```diff
@@ -172,6 +175,33 @@ func createOutlierDetectionConfig(upstream *v1.Upstream) (*envoy_config_cluster_
 	return api_conversion.ToEnvoyOutlierDetection(upstream.GetOutlierDetection()), nil
 }
 
+func convertDefaultSubset(defaultSubset *v1_options.Subset) *_struct.Struct {
+	if defaultSubset != nil {
```
Comment (kdorosh) [inline projects/gloo/pkg/translator/clusters.go:179]: nit, it's easier to read with the code flattened

e.g.
```golang
if defaultSubset == nil {
  return nil
}

// rest of logic here
```

Comment (jannes) [reply projects/gloo/pkg/translator/clusters.go]: thanks, i improved this

**Comment by jannes** [comment]
I have modified one test to make sure the defaults are correct and added one test to check the defaultSubset translation. Let me know if you think there should be more tests:)

File: projects/gloo/api/v1/options/subset_spec.proto [hunk 2]
```diff
@@ -7,9 +7,24 @@ option (extproto.equal_all) = true;
...
+  // single_host_per_subset is false by default
+  bool single_host_per_subset = 2;
+}
+
+message Subset {
+  map<string, string> values = 1;
+}
+
+enum FallbackPolicy {
+  // ANY_ENDOINT is the default
```
Comment (saiskee) [inline projects/gloo/api/v1/options/subset_spec.proto:26]: [super nit] `ANY_ENDPOINT`

File: projects/gloo/api/v1/options/subset_spec.proto [hunk 3]
```diff
@@ -7,9 +7,24 @@ option (extproto.equal_all) = true;
 option (extproto.hash_all) = true;
 
 message SubsetSpec {
-    repeated Selector selectors =1;
+  repeated Selector selectors = 1;
```
Comment (saiskee) [inline projects/gloo/api/v1/options/subset_spec.proto:10]: Can we include comments describing each of these fields and the expected behavior? Any comments directly above the field will pulled into our official documentation when you run `make generated-code -B`.

Comment (jannes) [reply projects/gloo/api/v1/options/subset_spec.proto:13]: fixed

**Comment by sam-heilbron** [comment]
Looks great! Letting it run through CI

**Comment by sam-heilbron** [comment]
@jannes Our codegen has changed recently, which is why the codegen check is failing. Can you run `make install-go-tools generated-code -B` and then re-push?

**Review by sam-heilbron — APPROVED** [review]

**Comment by jannes** [comment]
thanks for fixing the generated code and merging! :)

**Pull request merged**